include/deal.II/distributed/error_predictor.h

```diff
@@ -0,0 +1,242 @@
+     *
+     * In theory, the combination of hp::Refinement::predict_error() to predict
+     * how the error will change with  parallel::distributed::CellDataTransfer
+     * to transfer data across meshes forms the intended way to apply the
+     * prediction algorithm. However, since p4est specifies grid refinement in
+     * the parallel distributed case, it yields more reliable and trustworthy
+     * results when we determine the predicted errors during adaptation process.
+     * This is achieved with this this class, similar to
+     * parallel::distributed::SolutionTransfer and the aforementioned
+     * parallel::distributed::CellDataTransfer classes.
+     *
+     * The following code snippet demonstrates how to impose hp-adaptivity based
+     * on refinement history in a parallel distributed application:
+     * @code
+     * // [initialisation...]
+     * Vector<float> predicted_error_per_cell(triangulation.n_active_cells());
+     * for(unsigned int i = 0; i < triangulation.n_active_cells(); ++i)
+     *   predicted_error_per_cell[i] = std::numeric_limits<float>::infinity();
+     *
+     * // [during each refinement step...]
+     * // set h-adaptivity flags
+     * Vector<float> estimated_error_per_cell(triangulation.n_active_cells());
+     * KellyErrorEstimator<dim>::estimate(...);
+     * parallel::distributed::GridRefinement::
+     *   refine_and_coarsen_fixed_{number|fraction}(...);
+     *
+     * // set p-adaptivity flags
+     * hp::Refinement::p_adaptivity_from_reference(
+     *   hp_dof_handler,
+     *   estimated_error_per_cell,
+     *   predicted_error_per_cell
+     *   std::less<float>(),
+     *   std::less<float>());
+     * hp::Refinement::{choose|force}_p_over_h(hp_dof_handler);
+     *
+     * // perform adaptation and predict error for the subsequent adaptation
+     * parallel::distributed::ErrorPredictor<dim> predictor(hp_dof_handler);
+     * predictor.prepare_for_coarsening_and_refinement(estimated_error_per_cell);
+     * triangulation.execute_coarsening_and_refinement();
+     *
+     * // unpack predicted errors
+     * predicted_error_per_cell.reinit(triangulation.n_active_cells());
+     * predictor.unpack(predicted_error_per_cell);
+     * @endcode
+     *
+     * For serialization of predicted errors, use the
+     * parallel::distributed::CellDataTransfer class.
+     *
+     * @ingroup distributed
+     * @author Marc Fehling, 2019 - 2020
+     */
+    template <int dim, int spacedim = dim>
+    class ErrorPredictor
+    {
+    public:
+      /**
+       * Constructor.
+       *
+       * @param[in] dof The hp::DoFHandler on which all operations will
+       *   happen. At the time when this constructor is called, the
+       *   DoFHandler still points to the triangulation before the
+       *   refinement in question happens.
+       */
+      ErrorPredictor(const hp::DoFHandler<dim, spacedim> &dof);
+
+      /**
+       * Destructor.
+       */
+      ~ErrorPredictor() = default;
```

Comment:
Ah, but that's just five extra lines of code :-) Unless a class has virtual functions, there is no point in explicitly declaring a destructor that is `=default`.